src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Telemetry/VsTelemetryService.cs

```diff
@@ -1,20 +1,142 @@
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem;
 
 namespace Microsoft.VisualStudio.Telemetry
 {
     [Export(typeof(ITelemetryService))]
-    internal class VsTelemetryService : ITelemetryService
+    [Export(typeof(IVsTelemetryService))]
+    internal class VsTelemetryService : IVsTelemetryService, ITelemetryService
     {
+        private const string EventPrefix = "vs/projectsystem/managed/";
+        private const string PropertyPrefix = "VS.ProjectSystem.Managed.";
+        private readonly IUnconfiguredProjectCommonServices _unconfiguredProjectCommonServices;
+        private readonly object _lock =  new object();
+
+        private TelemetryEventCorrelation _telemetryEventCorrelation;
+        private bool _telemetryEventCorrelationInitialized;
+        private Guid _projectGuid;
+
+        [ImportingConstructor]
+        public VsTelemetryService(IUnconfiguredProjectCommonServices unconfiguredProjectCommonServices)
+        {
+            Requires.NotNull(unconfiguredProjectCommonServices, nameof(unconfiguredProjectCommonServices));
+
+            _unconfiguredProjectCommonServices = unconfiguredProjectCommonServices;
+        }
+
+        /// <summary>
+        /// Creates the Project Correlation asset which will be used to correlate all the telemetry events for this project
+        /// identified uniquely through the Project Guid.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> CreateProjectCorrelationAssetAsync()
+        {
+            var configurationGeneralProperties = await _unconfiguredProjectCommonServices
+                                                .ActiveConfiguredProjectProperties
+                                                .GetConfigurationGeneralPropertiesAsync()
+                                                .ConfigureAwait(false);
+
+            if (Guid.TryParse((string)await configurationGeneralProperties.ProjectGuid.GetValueAsync().ConfigureAwait(false), out Guid guid))
+            {
+                _projectGuid = guid;
+            }
+            else
+            {
+                return false;
+            }
+
+            var targetFrameworkProperty = await configurationGeneralProperties.TargetFramework.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+            var targetFrameworksProperty = await configurationGeneralProperties.TargetFrameworks.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+            var outputTypeProperty = await configurationGeneralProperties.OutputType.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+
+            lock (_lock)
+            {
+                if (!_telemetryEventCorrelationInitialized)
+                {
+                    var eventName = "Project";
+                    var projectProperties = new Dictionary<string, object>();
+
+                    if (!string.IsNullOrEmpty(targetFrameworksProperty))
```

Comment:
Hmm but the properties are not immutable. So imagine the first root event says the targetframework is net45 and the user changes it to net46 - now all future events will be based on net46 but if we correlate back, we'll think the project targets net45. In this case maybe the root event should only have the thing that makes it unique and all other information is useless because it depends on when that event was fired.